12/WAKU2-FILTER: Update (#119)
vacp2p/rfc-index: waku/standards/core/12/filter.md

  - Ebube Ud <ebube@status.im>
---

previous versions: [00](/waku/standards/core/12/previous-versions/00/filter.md)

**Protocol identifiers**:

- _filter-subscribe_: `/vac/waku/filter-subscribe/2.0.0-beta1`
- _filter-push_: `/vac/waku/filter-push/2.0.0-beta1`

---

## Abstract

This specification describes the `12/WAKU2-FILTER` protocol,

> which enables subscribing to messages that a peer receives.

Reviewer: Not your formulation, I know, but I think the original is a bit inexact. :)

```suggestion
which enables a client to subscribe to a subset of real-time messages from a Waku peer.
```

This is a more lightweight version of [11/WAKU2-RELAY](/waku/standards/core/11/relay.md),

> specifically designed for bandwidth restricted devices.

Reviewer: It's certainly not only for bandwidth-restricted devices

```suggestion
useful for bandwidth restricted devices.
```

> This is due to the fact that light-nodes subscribe to full-nodes and

Reviewer: ```suggestion
This is often used by lightweight nodes to subscribe to full Relay nodes and
```

Author: Does lightweight nodes refer to Waku nodes using lightpush? The light node terminology has been used in other RFCs, is this not the preferred term?

Author: @jm-clius As discussed prior, instead of using lightweight nodes, the suggested statement:
" This is often used by nodes with lower resource limits to subscribe to full Relay nodes and only receive the subset of messages they desire, based on content topic interest."

Reviewer: Thanks. Sounds better!

> only receive the messages they desire.

Reviewer: ```suggestion
only receive the subset of messages they desire, based on content topic interest.
```

## Motivation

Unlike the [13/WAKU2-STORE](/waku/standards/core/13/store.md) protocol
for historical messages, this protocol allows for native lower latency scenarios,
such as instant messaging.
It is thus complementary to it.

Strictly speaking, it is not just doing basic request-response, but
performs sender push based on receiver intent.
While this can be seen as a form of light publish/subscribe,
it is only used between two nodes in a direct fashion. Unlike the

> Gossip domain, this is meant for light nodes which put a premium on bandwidth.

Reviewer: ```suggestion
Gossip domain, this is suitable for light nodes which put a premium on bandwidth.
```

No gossiping takes place.

It is worth noting that a light node could get by with only using the
[13/WAKU2-STORE](/waku/standards/core/13/store.md) protocol to
query for a recent time window, provided it is acceptable to do frequent polling.

## Semantics

The key words “MUST”, “MUST NOT”, “REQUIRED”,
“SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and
“OPTIONAL” in this document are to be interpreted as described in [2119](https://www.ietf.org/rfc/rfc2119.txt).

### Content filtering

Content filtering is a way to do
[message-based filtering](https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern#Message_filtering).
Currently the only content filter being applied is on `contentTopic`.

> This corresponds to topics in [6/WAKU1](waku/standards/legacy/6/waku1.md).

Reviewer: ```suggestion
```
References to Waku 1 not really relevant anymore.

For example, the signature verification key, and
the hash of one's static IP address are unique for each user and hence count as PII.

### Design Requirements

The effectiveness and reliability of the content filtering service enabled by the
`WakuFilter` protocol rely on the _high availability_ of the full nodes
as the service providers.
To this end, full nodes must feature _high uptime_
(to persistently listen and capture the network messages)

> as well as _high Bandwidth_ (to provide timely message delivery to the light nodes).

Reviewer: Let's remove this entire section.
